src/main/java/org/truffleruby/language/constants/ReadConstantNode.java

```diff
@@ -35,9 +36,25 @@ public ReadConstantNode(RubyNode moduleNode, String name) {
     public Object execute(VirtualFrame frame) {
         final Object module = moduleNode.execute(frame);
 
-        final RubyConstant constant = lookupConstantNode.lookupConstant(frame, module, name);
+        final RubyConstant constant = getLookupConstantNode().lookupConstant(frame, module, name);
 
-        return getConstantNode.executeGetConstant(frame, module, name, constant, lookupConstantNode);
+        return getGetConstantNode().executeGetConstant(frame, module, name, constant, getLookupConstantNode());
+    }
+
+    private GetConstantNode getGetConstantNode() {
+        if (getConstantNode == null) {
+            CompilerDirectives.transferToInterpreterAndInvalidate();
+            getConstantNode = insert(GetConstantNode.create());
+        }
+        return getConstantNode;
+    }
+
+    private LookupConstantNode getLookupConstantNode() {
+        if (lookupConstantNode == null) {
+            CompilerDirectives.transferToInterpreterAndInvalidate();
+            lookupConstantNode = insert(LookupConstantNodeGen.create(false, false));
+        }
+        return lookupConstantNode;
```

Comment:
Same comment regarding including the `execute` call inside the lazy pattern method.

Reply:
true, but here it comes at the cost of some code duplication, as `executeGetConstant` also requires `lookupConstant`.

Reply:
Ah, right. I missed `executeGetConstant()` takes a LookupConstantNode as argument as well. Good as-is then.